### PR TITLE
Speed up SMILES writer by replacement of SSSR in SMILES writer with a bounded BFS

### PR DESCRIPTION
**Pull request by baoilleach**
To test whether a double bond is implicitly cis (and thus requires no stereobond symbols), the original code used SSSR based ring perception. In fact, it always triggered it, even where the bond was not in a ring. Doh!

The replacement code does not use the SSSR.

### CODE REVIEW

File: src/formats/smilesformat.cpp [hunk 1]
```diff
@@ -2364,6 +2364,57 @@ namespace OpenBabel {
...
   }
 
+  // Is this dbl bond in a ring that is of size <= IMPLICIT_CIS_RING_SIZE?
+  //
+  // A bounded BFS from one side of a bond trying to find the other side.
+  //   - The required queue is implemented using a std::vector for efficiency
+  //   - Note that items are never popped from the queue, I just move a pointer
+  //     along to mark the left hand side
+  //   - A potential improvement would be to BFS from both sides at the same time
+  static bool IsNotInRingOfSizeThatImpliesCis(OBMol *mol, OBBond *bond)
```
Comment (ghutchis) [inline src/formats/smilesformat.cpp]: I'd actually prefer if we can make this a general method, e.g., OBMolAtomBFSIter. Many of the calls to SSSR could be replaced by an iterator like this (e.g., aromaticity)

File: src/formats/smilesformat.cpp [hunk 2]
```diff
@@ -2364,6 +2364,57 @@ namespace OpenBabel {
...
+      OBAtom* curr = qatoms[qstart];
+      if (qstart == depthmarker) {
+        depth++;
+        depthmarker = qatoms.size();
+      }
+      qstart++;
+      if (seen.BitIsSet(curr->GetIdx()))
+        continue;
+      seen.SetBitOn(curr->GetIdx());
+      if (depth < IMPLICIT_CIS_RING_SIZE) {
```
Comment (ghutchis) [inline src/formats/smilesformat.cpp]: IMPLICIT_CIS_RING_SIZE should be a method parameter

File: test/testbindings.py [hunk 3]
```diff
@@ -88,6 +88,21 @@ def testAtomMapsAfterDeletion(self):
         mol.removeh()
         self.assertEqual(mol.write("smi", opt={"a":True}).rstrip(), "C[NH:2]")
 
+    def testImplicitCisDblBond(self):
```
Comment (ghutchis) [inline test/testbindings.py]: This came up in the coordinate generation discussions... what should the code do when presented with a SMILES with a trans indication in a small ring? Right now, we don't have any sort of stereo warning.

Comment (baoilleach) [reply test/testbindings.py]: Actually I almost did this. Well I did it, and then reverted, to be honest. But I can put it back. My only concern is that while this is efficient for a single atom, I think a different approach would be better if you are interested in identifying all atoms in smallish rings (I had a discussion on this with @johnmay).

But for sure, there's no harm in making the method available.

File: src/formats/smilesformat.cpp (hunk 1 above)
Comment (baoilleach) [reply src/formats/smilesformat.cpp]: Actually I almost did this. Well I did it, and then reverted, to be honest. But I can put it back. My only concern is that while this is efficient for a single atom, I think a different approach would be better if you are interested in identifying all atoms in smallish rings (I had a discussion on this with @johnmay).

But for sure, there's no harm in making the method available.

File: src/formats/smilesformat.cpp (hunk 2 above)
Comment (baoilleach) [reply src/formats/smilesformat.cpp]: Sure.

File: test/testbindings.py (hunk 3 above)
Comment (baoilleach) [reply test/testbindings.py]: Sure. It could be part of the PerceiveStereo.

**Comment by baoilleach** [comment]
I think these changes address your comments (apart from the comment about trans stereo in small rings).

**Comment by baoilleach** [comment]
Or did you have in mind an actual iterator?

**Comment by ghutchis** [comment]
I had a bounded iterator in mind, but I can take it from there. No reason to hold up this patch, except for the weird Travis fail.

**Comment by baoilleach** [comment]
Hmmm...seems to work fine on Windows (Travis doesn't run the Windows tests, but I've just done them here). I'll check it out on Linux...

**Comment by baoilleach** [comment]
After a reboot, the Windows tests did fail - this does happen occassionally where Windows refuses to believe the dlls have changed. Anyway, dumb mistake - now fixed.

**Pull request merged**